src/software_version.cr

```diff
@@ -0,0 +1,330 @@
+        # They are compared (numerical) and the algorithm only continues if they
+        # are equal.
+        ret = lnumber <=> rnumber
+        return ret unless ret == 0
+
+        # Move to the next position in both strings
+        lindex = new_lindex
+        rindex = new_rindex
+      end
+    end
+  end
+
+  # Helper method to read a sequence of digits from *string* starting at
+  # position *index* into an integer number.
+  # It returns the consumed number and index position.
+  private def consume_number(string : String, index : Int32)
+    number = 0
+    while (byte = string.byte_at?(index)) && byte.chr.ascii_number?
+      number *= 10
+      number += byte
+      index += 1
+    end
+    {number, index}
+  end
+
+  # Implements the pessimistic version constraint `~>`.
+  #
+  # A version matches *constraint* if is equal to the constraint comparing up
+  # to the constraint's second-to-last segment and the following segment is
+  # greater than the constraint's.
+  #
+  # ```
+  # require "software_version"
+  #
+  # SoftwareVersion.parse("1.0.0").matches_pessimistic_version_constraint?("1.0.0") # => true
+  # SoftwareVersion.parse("1.0.1").matches_pessimistic_version_constraint?("1.0.0") # => true
+  # SoftwareVersion.parse("1.1.0").matches_pessimistic_version_constraint?("1.0.0") # => false
+  # SoftwareVersion.parse("1.1.0").matches_pessimistic_version_constraint?("1.0")   # => true
+  # SoftwareVersion.parse("2.0.0").matches_pessimistic_version_constraint?("1.0")   # => false
+  # ```
+  def matches_pessimistic_version_constraint?(constraint : String) : Bool
+    constraint = self.class.parse(constraint).release.to_s
+
+    if last_period_index = constraint.rindex('.')
+      constraint_lead = constraint.[0...last_period_index]
+    else
+      constraint_lead = constraint
+    end
+    last_period_index = constraint_lead.bytesize
+
+    # Compare the leading part of the constraint up until the last period.
+    # If it doesn't match, the constraint is not fulfilled.
+    return false unless @version.starts_with?(constraint_lead)
+
+    # The character following the constraint lead can't be a number, otherwise
+    # `0.10` would match `0.1` because it starts with the same three characters
+    next_char = @version.byte_at?(last_period_index).try &.chr
+    return true unless next_char
+    return false if next_char.ascii_number?
+
+    # We've established that constraint is met up until the second-to-last
+    # segment.
+    # Now we only need to ensure that the last segment is actually bigger than
+    # the constraint so that `0.1` doesn't match `~> 0.2`.
+    # self >= constraint
+    constraint_number, _ = consume_number(constraint, last_period_index + 1)
+    own_number, _ = consume_number(@version, last_period_index + 1)
+
+    own_number >= constraint_number
+  end
+
+  # Custom hash implementation which produces the same hash for `a` and `b` when `a <=> b == 0`
+  def hash(hasher)
```

Comment:
An example would be welcome, along with type annotation if possible. Is there any advantage over `<=>` – performance?

Reply:
This is just the implementation of `Object#hash` and ensures that when `a <=> b == 0` both instances have the same hash.
here is no advantage over `<=>`. Both methods serve different methods.